[Mobile Payments]  IPP onboarding in Canada when WCPay and Stripe are both installed and activated

toupper opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #6802
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
When both WCPay and Stripe plugins are installed and active in a store based in a country that doesn't support Stripe (e.g. Canada), we used to show the onboarding view letting the user know just that "Stripe is not supported in Canada", which doesn't advise the best way to proceed (just to deactivate Stripe).

|<img src="https://user-images.githubusercontent.com/2472348/167425689-b43ee570-4f9c-49e8-b6ea-da21c39eec1d.jpg" width="375" caption="test">|
|:--:| 
|*Current Onboarding view without the changes of this PR*|


With this PR we create a new onboarding case when WCPay and Stripe are both installed and activated in a country that doesn't support Stripe, advising them to just deactivate Stripe so the payments can be collected with WCPay.

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
#### Prerequisites
- Store based in Canada
- WCPay and Stripe Gateway plugins installed and activated
- IPP in Canada feature is on in the app
#### Steps
- Go to Menu
- Tap on Settings
- Tap on In-Person Payments
- The onboarding view suggests deactivating Stripe to proceed (see screenshots)

### Screenshots
<!-- Include before and after images or gifs when appropriate. -->
|Admin|Shop Manager|
-|-
<img src="https://user-images.githubusercontent.com/1864060/169067239-f180d5fc-d3ed-4ab6-b36f-9d2da4eefb4d.png" width="375">|<img src="https://user-images.githubusercontent.com/1864060/169067256-2cc731aa-55ca-4b73-aa04-0d6a19e40a8f.png" width="375">

---
- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

toupper committed:
WIP: Add new state for Stripe not supported in a country, should be deactivated

toupper committed:
Add screen for the new onboarding state

toupper committed:
Enhance onboarding event

toupper committed:
Adapt UI to refactored event.

toupper committed:
Add unit test for both payment plugins installed at the same time in Canada

toupper committed:
Add documentation

toupper committed:
Refine comment

toupper committed:
Merge branch 'trunk' into issue/6802-stripe-not-supported-in-canada-onboarding

# Conflicts:
#	WooCommerce/WooCommerce.xcodeproj/project.pbxproj

toupper commented on Yosemite/Yosemite/Model/Enums/CardPresentPaymentsOnboardingState.swift:
```diff
@@ -77,6 +82,8 @@ extension CardPresentPaymentOnboardingState {
         case .selectPlugin:
             return "multiple_plugins_installed"
+        case .deactivateStripe:
+            return "stripe_should_be_deactivated"
```
I couldn't find a source for these tracking strings other than this PR https://github.com/woocommerce/woocommerce-android/pull/4562 . Is there a table that I can follow/update for this new string? @malinajirka

malinajirka replied:
[Here are all the onboarding errors with the corresponding events](https://github.com/woocommerce/woocommerce-android/blob/1833f7994d21851e26c6dc6164000f8fa1033e59/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/cardreader/CardReaderTracker.kt#L121) we track on Android.

It's also a good practice to update the google sheet table - I'll share a link in a dm.

---------------------

I don't want to block this PR and I might be missing some context, so feel free to ignore this comment. However, I'm not sure I understand why we force the user to disable the Stripe plugin in this scenario? Could we perhaps just tell the app to use wcpay endpoints and completely ignore the fact that Stripe plugin is installed? AFAIK there isn't any conflict between the plugins so asking the user to disable an unsupported plugin feels unnecessary. Wdyt?

toupper replied:
Thanks for the information @malinajirka!!

> I don't want to block this PR and I might be missing some context, so feel free to ignore this comment. However, I'm not sure I understand why we force the user to disable the Stripe plugin in this scenario? Could we perhaps just tell the app to use wcpay endpoints and completely ignore the fact that Stripe plugin is installed? AFAIK there isn't any conflict between the plugins so asking the user to disable an unsupported plugin feels unnecessary. Wdyt?

@joshheald already thought about that option in the parent issue [here](https://github.com/woocommerce/woocommerce-ios/issues/6802): 

> Alternatively, since we don't support Stripe in Canada but do support WCPay, we could ignore the fact that both are installed and continue with WCPay regardless, so avoid showing an onboarding failure. I don't think that would be a good idea with our current design, since we are likely to support either (but not both) one day soon, so a merchant who keeps both plugins for international markets may order a reader now, and find it unusable in a few months time.

Also pdfdoF-FO-p2 (even if we didn't discuss this exact scenario there)

@malinajirka @joshheald @ctarda @joe-keenan  It would be great if we could align Android and iOS here so we go with the same approach :) What do you think?

malinajirka replied:
Thanks so much for the context 🙇 !

> we could ignore the fact that both are installed and continue with WCPay regardless, so avoid showing an onboarding failure. I don't think that would be a good idea with our current design, since we are likely to support either (but not both) one day soon

I believe this might not be a valid concern anymore. We plan to let the user choose which plugin they want to use when the app detects both plugins are installed: More context on pdfdoF-FO-p2#comment-1537.


I think we might want to consider the following:
1. Use WCPay in Canada no matter if Stripe is installed or not.
2. Implement support for letting the user choose which plugin they want to use (show it just in the US until we support Stripe in Canada).


P.S. If we decide to merge this PR, I'd consider removing `stripe_` from the event and using generic `plugin_should_be_deactivated`. 
Including the plugin name in the events bit us in the past (eg. we need to have two separate events for `stripe_not_setup` and `wcpay_not_setup` instead of having a single event `plugin_not_setup` with a plugin_type property). Wdyt?

malinajirka replied:
Thinking about this more, perhaps we could
1. Merge this PR.
2. Mirror it on Android.
3. Implement support for letting the user choose + remove the screen introduced in this PR.

We'd avoid the discussion paralysis and implementing it on Android should be pretty straightforward.

Wdyt?

toupper replied:
Thanks @malinajirka for your feedback!

> P.S. If we decide to merge this PR, I'd consider removing stripe_ from the event and using generic plugin_should_be_deactivated.
> Including the plugin name in the events bit us in the past (eg. we need to have two separate events for stripe_not_setup and wcpay_not_setup instead of having a single event plugin_not_setup with a plugin_type property). Wdyt?

My first approach was exactly that, to create a generic event that could handle any kind of plugin, something like:
`pluginShouldBeDeactivated(plugin: CardPresentPaymentsPlugin)`
Unfortunately making it so generic made the implementation of the view a bit too complex:
The view currently has the title "In-Person Payments are processed through WooCommerce Payments." If we make the case generic we would have to make the view title generic as well and pass the plugin that should remain as parameter, otherwise we could have the strange case where we pass the WCPay as parameter `pluginShouldBeDeactivated(plugin: .wcPay)`, ending up in something like:
Title: "In-Person Payments are processed through WooCommerce Payments."
Message: "Please deactivate WooCommerce Payments."
What do you think? I see four options here:

- Leave it as it is
- Make the title more generic, something like "This plugin needs to be deactivated to collect payments". Unfortunately this doesn't bring as much context as the current one.
- Make the case generic, passing only the deactivating plugin `pluginShouldBeDeactivated(plugin: CardPresentPaymentsPlugin)` causing the edge inconsistent case shown above.
- Make the case double generic passing the plugin that should be kept, something like: `pluginShouldBeDeactivated(plugin: CardPresentPaymentsPlugin, shouldBeKept: CardPresentPaymentsPlugin)`
@joshheald 

> Thinking about this more, perhaps we could
> 
> Merge this PR.
> Mirror it on Android.
> Implement support for letting the user choose + remove the screen introduced in this PR.

I agree with that 👍

malinajirka replied:
>  otherwise we could have the strange case where we pass the WCPay as parameter pluginShouldBeDeactivated(plugin: .wcPay), ending up in something like

What if we simply threw "NotImplementException" in this case - it's a case that should never happen and if it happens we want to know about it ASAP and fix it right? Wdyt?

toupper replied:
> What if we simply threw "NotImplementException" in this case - it's a case that should never happen and if it happens we want to know about it ASAP and fix it right? Wdyt?

That makes sense! I Implemented something similar in https://github.com/woocommerce/woocommerce-ios/pull/6892/commits/dd0ad80cdf9e33354d1c958a55f34d7e67a77749, checking that the plugin to be deactivated is stripe before showing the view. Otherwise, it goes to the default onboarding view `InPersonPaymentsUnavailable`, that should never happen for the case we have here.
Thanks a lot @malinajirka for the feedback here 🚀

toupper committed:
Fix text

toupper committed:
Make onboarding state case more generic

toupper committed:
Fix unit tests

jaclync commented on WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person Payments/Onboarding Errors/InPersonPaymentsDeactivateStripeAdminView.swift:
```diff
@@ -0,0 +1,85 @@
+
+struct InPersonPaymentsDeactivateStripeAdmin: View {
+    let onRefresh: () -> Void
+    @State var presentedSetupURL: URL? = nil
```
nit: could this state var be private?
```suggestion
    @State private var presentedSetupURL: URL? = nil
```

jaclync commented on WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person Payments/Onboarding Errors/InPersonPaymentsDeactivateStripeAdminView.swift:
```diff
@@ -0,0 +1,85 @@
+    @State var presentedSetupURL: URL? = nil
+    @Environment(\.verticalSizeClass) var verticalSizeClass
+
+    var isCompact: Bool {
```
nit: is this used somewhere?

jaclync commented on WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person Payments/Onboarding Errors/InPersonPaymentsDeactivateStripeAdminView.swift:
```diff
@@ -0,0 +1,85 @@
+private enum Localization {
+    static let title = NSLocalizedString(
+        "In-Person Payments are processed through WooCommerce Payments.",
+        comment: "Title for the error screen when there is more than one plugin active" +
+        "and the Stripe plugin should be deactivated"
```
nit: maybe a space between these two strings?

jaclync commented on WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person Payments/Onboarding Errors/InPersonPaymentsDeactivateStripeAdminView.swift:
```diff
@@ -0,0 +1,85 @@
+                message: Localization.message,
+                image: InPersonPaymentsOnboardingError.ImageInfo(
+                    image: .paymentErrorImage,
+                    height: 108.0
```
nit: 
```suggestion
                    height: Constants.height
```

jaclync commented on WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person Payments/Onboarding Errors/InPersonPaymentsDeactivateStripeShopManagerView.swift:
```diff
@@ -0,0 +1,69 @@
+import SwiftUI
+import Yosemite
+
+struct InPersonPaymentsDeactivateStripeShopManager: View {
```
💭 it looks like the only difference between this view and `InPersonPaymentsDeactivateStripeAdmin` is the button to manage plugins? what do you think if we pass in a boolean that indicates whether to show this button with `viewModel.userIsAdministrator` in `InPersonPaymentsViewController`?

jaclync commented on WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person Payments/Onboarding Errors/InPersonPaymentsDeactivateStripeAdminView.swift:
```diff
@@ -0,0 +1,85 @@
+import SwiftUI
+import Yosemite
+
+struct InPersonPaymentsDeactivateStripeAdmin: View {
```
nit: what do you think of naming this struct with `View` at the end? If I just read this name I'd not have guessed this is a view. I understand that our codebase has similar naming for SwiftUI views, just wanted to mention since I saw we're good with naming SwiftUI views as `*View` in p1651699927182139/1651641104.110959-slack-CGPNUU63E (at least from the number of 👍 )

jaclync commented on WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person Payments/Onboarding Errors/InPersonPaymentsDeactivateStripeAdminView.swift:
```diff
@@ -0,0 +1,85 @@
+        .safariSheet(url: $presentedSetupURL, onDismiss: onRefresh)
+    }
+
+    var setupURL: URL? {
```
nit: could this be private?

jaclync reviewed (approved):
Works as expected! Just some non-blocking nits and suggestions 🚢 

Tested with an admin then deactivating Stripe to see card reader settings, shop manager, and a store with an unsupported country ✅

toupper committed:
Make property private

toupper committed:
Remove not needed property

toupper committed:
Add space between strings

toupper committed:
Use constant for the height

toupper committed:
Rename to make it more meaningful

toupper committed:
Make properties private

toupper committed:
Remove not needed properties

toupper committed:
Collapse views into one with a boolean value for customization

toupper committed:
Remove debug logic

toupper committed:
Improve strings

toupper commented:
@jaclync Thanks for the review! I addressed all your comments and tweaked the strings according to the comment from @joe-keenan [here](https://github.com/woocommerce/woocommerce-ios/issues/6802#issuecomment-1134379313)
